### DISCUSSION TITLE
How to?: Query two postgreSQL TABLES, filter to create and update a single combined TABLE

### DISCUSSION THREAD

**Discussion by PaulRossAclear** [Budibase/budibase#8734] [Help]
Edit: Edited the Topic and comments to clarify the distinction between table and database. 
We have been able to connect two postgres data bases but only able to query one database at a time using a screen tool. Need assistance with query multiple database and filter for specific results.  Once the desired records form each database have been identified, would like to create a combined database that will also be a postgres database.  Most helpful then would be to create a unique identifier for each combined record in the new database preferably a randomly generated rather than sequential.  This can be manually managed.  eg. user queries both databases, selects desired results and commits to new database with unique randomly generated serial so that in the new database these records are tied together.  It may be helpful to have the source databases appended with this unique id so that they can be identified as having been "matched" and can be found in the new database.  

More background:
The concept is to tie multiple records together for a subject.  ie. for analysis of real estate. A home may be sold several times and those listings may be recorded on a MLS service and a tax records database.  The MLS records and the tax records may not match and often the tax records lag behind the MLS, so depending on the date of collection the data may not yet be posted.  The MLS also may be recorded as sold when in fact it did not.  The transaction may be worth analysis but it should be identified correctly.  They may also have been listed for rent at various times so these records need to be recorded and tied into a search.  More functionality and features are planned but this most basic step of creating the database of tied records, we feel is the most critical element.

**Comment by melohagan** [comment]
Hey @PaulRossAclear 

If the tables are in different databases, then naturally there's not going to be a convenient way to join these tables, as far as I'm aware.

What you may need to do is form "soft" relationships where you have an ID field in one table that matches the primary key of the other, and then you can use filters on your data providers to get related rows. 
Still I believe it may not be good practice to have relationships between tables in different databases.

**Reply by PaulRossAclear** [reply]
Please forgive me if I get some of the terminology incorrect.  
I may have caused some confusion by referring to Tables as Databases.  SO what we actually want is to create a combined table from two or more separate tables.
I am reading through "How to Integrate Multiple Databases in a Web App"  https://budibase.com/blog/data/how-to-integrate-multiple-databases/ 

We have completed step 1
 https://budibase.com/blog/data/how-to-integrate-multiple-databases/#:~:text=1.%20Connect%20your%20data 

Except both our tables are in postgreSQL and hosted on Amazon AWS RDS 

I have managed to use the code in step 2 and edit to match my database descriptions. see pic attached
![image](https://user-images.githubusercontent.com/118623467/203124717-a98ab5b6-61a0-488e-82ce-5ca76f1f0e84.png)

was able to create 2 new tables using the Budibase Query
See screen shot of postgres PgAdmin
![image](https://user-images.githubusercontent.com/118623467/203127841-5fb46d13-7edd-4821-8079-d54f7d4145a4.png)

Combined_customers_table is now a copy of mytable and combined_mls_tx is a copy of mytable1

For my case I do not want to update the source tables mytable and mytable1 rather, I would like to make a combined table which can be search based on a few different criteria then I can create a set of matched records and then merge those into a new combined table with a unique identifier assigned to that set.
I'll refer to my steps below as sub-steps to avoid any confusion with the steps referred to in the referenced documents
So summarizing my comments I see this as two separate steps.  sub-step 1 Create a combined table.  sub-step 2 search that new table sub-step 3 create a separate new table containing these sets of matched records with a unique ID which is the same for all records in the identified set.

I suppose sub-step 3 could be modified to create a unique ID for the set and just update those fields with the "Unique set" ID

I am presently stuck at step 2 of how to actually create a combined table.  For reference each of my records contains approximately 75 columns so the combined table will have approximately 150 columns.

**Reply by melohagan** [reply]
Is there any reason you couldn't create the combined table ahead of time, and insert rows into it as needed, rather than creating it via a query in Budibase?
If you just needed to combine the data for the purposes of viewing it, you could use a query with a JOIN?

**Reply by PaulERayburn** [reply]
On further consideration, I think the tables may not need to be merged.  

The goal is to search multiple tables for various fields criteria to identify records related to the same property.  For example, we will have multiple sales of a given property in both MLS and Tax Records.  (Not all properties have PID's or PUD's so we want to create our own). 

The idea is to manually search and validate these records and then identify them all with a common ID so that for future analysis the validation is already complete and anytime a property is identified in a subsequent search which has the ID assigned is known to be validated.  From this stage we would need to be able to add those to another table or export for further analysis for a specific assignment.  In our case, they will be analyzed using R via Rstudio so a "plug" into this table(s) may be possible, minimizing exporting and enhancing the ability to update/ revise searches.  

Some of the complicating factors are that we need to identify the source of the data. 

  In my case, I have chosen to add the prefix TX to all tax records field names which would allow the user to identify that source.  The MLS records have a unique MLS# tied to every record with the exception that the active records will become duplicates in subsequent searches which are added to the database as they will have the initial "Active" status then "Sold" or "Withdrawn etc.".  We could add the prefix ML to each of those fields for consistency.  We will have additional sources of information which will be stored in another table(s) to be queried and identified to the "property" in the same process as described.  

So we do not want to combine or merge fields. For this reason, the suggestion of pre combining the tables may not make sense.  I have tested this and of course, it is possible in a CSV file however very cumbersome. (pasting the field names to the end of the row of one source and then pasting the data below the last row of cells for the initial data set and so on for additional records.  Excel may offer some automation to do this so I may need to dig into that a bit further but I am hoping this can be managed with a CRUD/APP interface.

One other consideration I have is to avoid confusion subsequent searches may find new records to be added to the set.  We need to determine whether to add the prior ID to the new record or create a whole new ID for the entire new set.  That could be handled either way whichever is easier.

**Reply by PaulERayburn** [reply]
I have attempted to query two tables in BB using forms, tables and field filters as an option.  Unless there is another method there is some redundant steps such as entering search fields twice.  The next step would be to assign the common unique ID to a set. Then find a way to export or access the tables with the new ID's.  Heres a screen shot of where I am at.  

![image](https://user-images.githubusercontent.com/117835153/203363080-63ecf449-dac8-40c4-a38d-05115f21fdc4.png)

![image](https://user-images.githubusercontent.com/117835153/203363201-af14b640-674b-415e-8588-8f2449591341.png)

**Reply by PaulERayburn** [reply]
So we're still working through this and I'm just trying random tests.  I have no clue how this is expected to work at this stage but I tried creating a button to update rows with a unique ID created with a javascript.  But it just crashes the browser with a memory when attempting preview 

![image](https://user-images.githubusercontent.com/117835153/203718508-d2c6914f-3931-4794-a45e-34dfe65f551f.png)

and heres the java script for unique id which appears to function in a javascript editor 
Here's a jsfiddle to demonstrate: http://jsfiddle.net/wSQBx/

![image](https://user-images.githubusercontent.com/117835153/203719972-2feffe35-5937-4736-beab-0834218e6d66.png)


here's the code from this  https://stackoverflow.com/questions/10726909/random-alpha-numeric-string-in-javascript

function randomString(length, chars) {
    var result = '';
    for (var i = length; i > 0; --i) result += chars[Math.round(Math.random() * (chars.length - 1))];
    return result;
}
document.write(randomString(32, '0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ'));

I'm shooting in the dark here.  Any ideas?

**Reply by melohagan** [reply]
Hey @PaulRossAclear 

Just to double check, your JavaScript binding is returning the random string, e.g.

```js
function randomString(length, chars) {
    var result = '';
    for (var i = length; i > 0; --i) result += chars[Math.round(Math.random() * (chars.length - 1))];
    return result;
}
return randomString(32, '0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ');
```

**Comment by PaulERayburn** [comment]
Yes, I verified the code will produce a random number string. My goal is to
have it populate the fields of the selected rows but it is just crashing
the program.

**Reply by melohagan** [reply]
Ah I just remembered that `Math.random` will cause Budibase to crash. Here's a related discussion: https://github.com/Budibase/budibase/discussions/5203

**Reply by PaulERayburn** [reply]
I just tried that code

function uuidv4() {
return ([1e7]+-1e3+-4e3+-8e3+-1e11).replace(/[018]/g, c =>
(c ^ crypto.getRandomValues(new Uint8Array(1))[0] & 15 >> c / 4).toString(16)
);
}

return uuidv4();

and it does not appear to function in a javascript but that is where my lack of coding skills is an issue. Is there something missing? Do I need to add something?
So it doesnt appear to function in a javascript editior
[image](https://user-images.githubusercontent.com/117835153/203804668-b991a7b6-cb4c-4a4d-9cc9-7b895e5272f4.png)

and when I try it in place of the math.random it still crashes BudiBase running in chrome. I have not tried other browsers.
[image](https://user-images.githubusercontent.com/117835153/203802274-5a5939d0-7366-42e0-b2ef-9f6caa8ca1c4.png)

**Reply by PaulERayburn** [reply]
Apologies, I see I responded to the wrong thread so I'm just going to post an image of your response here so we don't lose this thread.
![image](https://user-images.githubusercontent.com/117835153/203818817-0b00adbe-18cd-41c2-8a8f-3d8ccd962963.png)

So assuming we can get the random number generated, I am still not clear on how I can get that into the table(s).
For clarity, that needs to be one random number generated and then added manually or automatically into the field columns for the selected records(rows).
Thanks for your patience with me on this.

**Reply by melohagan** [reply]
You could possible try adding a formula column, that uses that code as a JavaScript binding value: https://docs.budibase.com/docs/formula

**Reply by PaulERayburn** [reply]
I don't understand how that would solve the issue of creating an id to tie together multiple selected rows.  This solution appears to only create a unique ID for every instance or row which would not solve anything.
Maybe I am misunderstanding.  
Thanks again for your assistance.

**Comment by drearystate** [comment]
If you are trying to search 2 databases at once you can use a form insert dataprovider and data and within that data provider and data have a second data provider and data (2nd database to search/filter through. the form can use form fields to filter results with both. I do this with looking for vehicles, I use 1 database to filter the other. When binding I use "like" instead of "equals" to help with spelling errors from others as well simple syntax errors.

**Reply by PaulERayburn** [reply]
So to be clear when we say database I think we are referencing tables. These tables could be one each in a separate database or multiple tables in one database.   But if you check my setup I think it was close to what you are mentioning.  
![image](https://user-images.githubusercontent.com/117835153/204721871-c4e58567-612e-4559-a6ae-2bb4320f40a2.png)

I was using Starts with for PID and then Like for address. I could not get my searches to filter both tables and had to filter each table separately.  Then the only option I could get to work was export CSV for each table. I would then need to either manually upload these sets to a new database table or create some kind of automation.

**Reply by drearystate** [reply]
Yes, and every field within that form can be used to filter any child database from the parent as well as the parent DB/table (data provider).

**Reply by drearystate** [reply]
In form, outside of the data providers at the top of the hierarchy put your filtering fields there. This is how mine looks. i added the filtering fields to the top of the 2nd data provider to filter everything below it. I know you cant see it but theres about 3 other data providers they filter below those.
![image](https://user-images.githubusercontent.com/20757700/204957164-0f1a559a-b295-42d7-adda-190b603dcf36.png)

**Comment by drearystate** [comment]
If you are wanting to merge the 2 databases then within the form mentioned above have action buttons alongside your tables, have the action fill out empty form fields on the screen (hide them if needed) and 1 final button that takes current fields and creates new database. It becomes a 3 click solution but it at least merges your databases.

**Reply by PaulERayburn** [reply]
That sounds promising.  I'd be OK with the extra 3 click solution.  Do you think it's possible to update the source database(s) with the Parent Child ID for identified sets?  I think in this manner once sets have been identified, subsequent searches of thise databases would reveal that sets/ matches have been made and can be easily identified.  If the source database/ tables did not store those ID's the new database/tables with the "sets" would have to also be referenced.  There may be an easy way and I may be over/under thinking that.

**Reply by drearystate** [reply]
Yes, I would add a new column to each DB that when you click the first button on each DB item it puts the same id in each so that they would be identical.
